File: Solution/0033.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int s = 0;
        int e = nums.size() - 1;
        
        while (s <= e) {
            int m = (s + e) / 2;
            
            if (nums[m] == target) return m;
            else if (nums[m] > target) {
                if (target < nums[s] && nums[m] >= nums[s]) s = m + 1;
                else e = m - 1;
            } else {
                if (target > nums[e] && nums[m] < nums[s]) e = m - 1;
                else s = m + 1;
            }
        }
        return -1;
    }
};

File: Solution/0033.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        start, end = 0, len(nums)-1
        
        while(start <= end):
            mid = (start + end) // 2
        
            if target == nums[mid]:
                return mid
            
            if nums[mid] < nums[start]:
                if (nums[mid] < target and target <= nums[end]):
                    start = mid + 1
                else:
                    end = mid - 1
                
            else:
                if (target < nums[mid] and nums[start] <= target):
                    end = mid -1
                else:
                    start = mid + 1
        
        return -1
                